### PR TITLE
add support for NIP-51 mute lists

### PR DESCRIPTION
**Pull request by bryanmontz**
Implements the first NIP-51 list event: a kind 10000 MuteListEvent.
https://github.com/nostr-protocol/nips/blob/master/51.md#standard-lists

Other list types will follow in a subsequent PR.

### CODE REVIEW

File: Sources/NostrSDK/Events/MuteListEvent.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        allValues(forTagName: .pubkey) ?? []
+    }
+    
+    /// The publicly muted event ids (threads).
+    public var eventIds: [String] {
+        allValues(forTagName: .event) ?? []
+    }
+    
+    /// The publicly muted hashtags.
+    public var hashtags: [String] {
```
Comment (tyiu) [inline Sources/NostrSDK/Events/MuteListEvent.swift:38]: I'm surprised that the compiler didn't complain about `hashtags` being declared twice due to this class extending `HashtagInterpreting`.

On a related note, we can update `HashtagInterpreting` and a few other classes to use the `allValues` helper. If you didn't want to add noise and bloat to this PR, we can do that in a separate PR as a small refactor.

File: Sources/NostrSDK/Events/MuteListEvent.swift [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    override init(kind: EventKind, content: String, tags: [Tag] = [], createdAt: Int64 = Int64(Date.now.timeIntervalSince1970), signedBy keypair: Keypair) throws {
+        try super.init(kind: kind, content: content, tags: tags, createdAt: createdAt, signedBy: keypair)
+    }
+    
+    init(content: String, tags: [Tag] = [], createdAt: Int64 = Int64(Date.now.timeIntervalSince1970), signedBy keypair: Keypair) throws {
+        try super.init(kind: .muteList, content: content, tags: tags, createdAt: createdAt, signedBy: keypair)
+    }
+    
+    /// The publicly muted public keys (authors).
+    public var pubkeys: [String] {
```
Comment (tyiu) [inline Sources/NostrSDK/Events/MuteListEvent.swift:28]: Wondering if it makes sense to have something like a `PubkeyTagInterpreting` as there's a few Nostr events that do it.

However, one issue that I see with refactoring in this way is that we lose some of the finer-grained documentation on the kind level.

File: Sources/NostrSDK/Events/MuteListEvent.swift [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        allValues(forTagName: .pubkey) ?? []
+    }
+    
+    /// The publicly muted event ids (threads).
+    public var eventIds: [String] {
```
Comment (tyiu) [inline Sources/NostrSDK/Events/MuteListEvent.swift:33]: Same thought here.

**Review by tyiu — COMMENTED** [review]
Looks great overall! Left some comments.

File: Sources/NostrSDK/Events/NostrEvent.swift [hunk 4]
```diff
@@ -99,4 +99,11 @@ public class NostrEvent: Codable {
     public func valueForRawTagName(_ tagName: String) -> String? {
```
[the remark this answers is not in the document]
Comment (bryanmontz) [reply Sources/NostrSDK/Events/NostrEvent.swift:99]: Agree. I changed the documentation, but I think this warrants the function names including "first". Thanks.

File: Sources/NostrSDK/Events/NostrEvent.swift [hunk 5]
```diff
@@ -99,4 +99,11 @@ public class NostrEvent: Codable {
...
         tags.first(where: { $0.name == tagName })?.value
     }
+    
+    /// All tags with the provided name.
+    /// - Parameter tag: The tag name to filter.
+    /// - Returns: The values associated with the tags of the provided name.
+    public func allValues(forTagName tag: TagName) -> [String]? {
```
[the remark this answers is not in the document]
Comment (bryanmontz) [reply Sources/NostrSDK/Events/NostrEvent.swift:106]: Yeah, in a separate PR I'll update all occurrences where this is applicable.

File: Sources/NostrSDK/Events/MuteListEvent.swift (hunk 1 above)
Comment (bryanmontz) [reply Sources/NostrSDK/Events/MuteListEvent.swift:38]: In a protocol extension, the function in the extension is just the "default" implementation and can be overridden by conformers.

Yep, coming in a separate PR.

Comment (bryanmontz) [reply Sources/NostrSDK/Events/MuteListEvent.swift:38]: I'll remove this function's implementation though, since it's just duplicating the one in the protocol.

File: Sources/NostrSDK/Events/MuteListEvent.swift (hunk 2 above)
Comment (bryanmontz) [reply Sources/NostrSDK/Events/MuteListEvent.swift:28]: Agree. It would be funny if you could override *just* the documentation. I think it's a good idea to move to a protocol for interpreting, but I'll do this in a next PR for nothing but tags improvements.

File: Sources/NostrSDK/Events/MuteListEvent.swift (hunk 3 above)
Comment (bryanmontz) [reply Sources/NostrSDK/Events/MuteListEvent.swift:33]: Yep, same.

File: Sources/NostrSDK/EventCreating.swift [hunk 6]
```diff
@@ -226,6 +226,69 @@ public extension EventCreating {
         return try ReportEvent(content: additionalInformation, tags: tags, signedBy: keypair)
     }
     
+    /// Creates a ``MuteListEvent`` (kind 10000) containing things the user doesn't want to see in their feeds. Mute list items be publicly visible or private.
+    /// - Parameters:
+    ///   - publiclyMutedPubkeys: Pubkeys to mute.
+    ///   - secretlyMutedPubkeys: Pubkeys to secretly mute.
```
[the remark this answers is not in the document]
Comment (bryanmontz) [reply Sources/NostrSDK/EventCreating.swift:232]: Yeah I think "private" is probably better. I was thinking about it from the client application's call site, but better to follow the spec we're implementing here. Will change.

**Comment by bryanmontz** [comment]
@tyiu Feedback addressed!

**Review by tyiu — APPROVED** [review]
